File: assets/hw/template/autograder/hw_release_sln/R/questions/ch3_ex14.R
# Chapter 3, Exercise 14.

suppressPackageStartupMessages({
  library(logging)
})

#' The width and height of exported plots.
FIGSIZE = c(1000, 1000)


#' Loads required data for exercise.
#'
#' @param data_path location of the data.
#' @return a data.frame containing the exercise data.
LoadData <- function(data_path ="../data/ch3_q14.csv", verbose=TRUE) {
  stopifnot(file.exists(data_path))
  if (verbose)
    loginfo(sprintf("\tLoading %s", data_path))
  sim_data <- read.csv(data_path)
  return (sim_data)
}


#' Using the cor() function, estimate the Pearson correlation
#' coefficient.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A numeric corresponding to the Pearson correlation.
PartB1 <- function(data_frame) {
  # YOUR CODE STARTS HERE
  est_cor <- with(data_frame, cor(x1, x2))
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (est_cor)
}


#' Return a scatter plot of x1 vs x2. Comment on what you see.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A matrix containing the pairwise correlation estimates.
PartB2 <- function(data_frame) {
  # YOUR CODE STARTS HERE
  figure <- with(data_frame, plot(x1 ~ x2))
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (figure)
}


#' Using the data, fit a least squares regression to predict y using
#' x1 and x2. Describe the results obtained.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A summary.lm object containing the summary of the regression.
PartC <- function(data_frame) {
  # YOUR CODE STARTS HERE
  fit_1 <- lm(y ~ x1 + x2, data=data_frame)
  lm_summary <- summary(fit_1)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (lm_summary)
}


#' Now fit a least squares regression to predict y using only x1.
#' Comment on your results.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A summary.lm object containing the summary of the regression.
PartD <- function(data_frame) {
  # YOUR CODE STARTS HERE
  fit_2 <- lm(y ~ x1, data=data_frame)
  lm_summary <- summary(fit_2)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE

  return (lm_summary)
}


#' Now fit a least squares regression to predict y using only x1.
#' Comment on your results.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A summary.lm object containing the summary of the regression.
PartE <- function(data_frame) {
  # YOUR CODE STARTS HERE
  fit_3 <- lm(y ~ x2, data=data_frame)
  lm_summary <- summary(fit_3)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (lm_summary)
}


#' Perform a multiple linear regression with y as the response and
#' x1, x2 as the predictors on the updated data. Return
#' the results using the summary() function Comment on the output.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A summary.lm object containing the summary of the regression.
PartG1 <- function(data_frame) {
  extra_obs = c(0.1, 0.8, 6)
  updated_data_frame = rbind(data_frame, extra_obs)
  # YOUR CODE STARTS HERE
  fit_1 <- lm(y ~ x1 + x2, data=updated_data_frame)
  lm_summary <- summary(fit_1)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (lm_summary)
}


#' Perform a multiple linear regression with y as the response and
#' x1 as the predictors on the updated data. Return
#' the results using the summary() function Comment on the output.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A summary.lm object containing the summary of the regression.
PartG2 <- function(data_frame) {
  extra_obs = c(0.1, 0.8, 6)
  updated_data_frame = rbind(data_frame, extra_obs)
  # YOUR CODE STARTS HERE
  fit_2 <- lm(y ~ x1, data=updated_data_frame)
  lm_summary <- summary(fit_2)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (lm_summary)
}


#' Perform a multiple linear regression with y as the response and
#' x2 as the predictors on the updated data. Return
#' the results using the summary() function Comment on the output.
#'
#' @param data_frame a data.frame containing the Simulation data.
#' @return A summary.lm object containing the summary of the regression.
PartG3 <- function(data_frame) {
  extra_obs = c(0.1, 0.8, 6)
  updated_data_frame = rbind(data_frame, extra_obs)
  # YOUR CODE STARTS HERE
  fit_3 <- lm(y ~ x2, data=updated_data_frame)
  lm_summary <- summary(fit_3)
  # stop("NotImplementedError")
  # YOUR CODE ENDS HERE
  return (lm_summary)
}

#' Wrapper function for Chapter 3, Exercise 14.
Ch3Ex14 <- function() {
  loginfo("Chapter 3, Exercise 14.")
  sim_data = LoadData()

  loginfo("\tPart b.")
  corr = PartB1(sim_data)
  sink("../results/R/ch3_ex14_part_b1.csv")
  print(corr)
  sink()

  png("../results/R/ch3_ex14_part_b2.png", width=FIGSIZE[1], height=FIGSIZE[2])
  PartB2(sim_data)
  dev.off()

  loginfo("\tPart c.")
  ols = PartC(sim_data)
  sink("../results/R/ch3_ex14_part_c.csv")
  print(ols)
  sink()

  loginfo("\tPart d.")
  ols = PartD(sim_data)
  sink("../results/R/ch3_ex14_part_d.csv")
  print(ols)
  sink()

  loginfo("\tPart e.")
  ols = PartE(sim_data)
  sink("../results/R/ch3_ex14_part_e.csv")
  print(ols)
  sink()

  loginfo("\tPart g.")
  ols = PartG1(sim_data)
  sink("../results/R/ch3_ex14_part_g1.csv")
  print(ols)
  sink()

  ols = PartG2(sim_data)
  sink("../results/R/ch3_ex14_part_g2.csv")
  print(ols)
  sink()

  ols = PartG3(sim_data)
  sink("../results/R/ch3_ex14_part_g3.csv")
  print(ols)
  sink()
}

File: assets/hw/template/autograder/hw_release_sln/python/questions/ch3_ex14.py
import os
from typing import Text, Tuple

from absl import logging
import matplotlib
import matplotlib.pyplot as plt
import pandas as pd
from scipy.stats import pearsonr
import statsmodels
import statsmodels.api as sm
from statsmodels.formula.api import ols

class Simulation(object):
  """Chapter 3, Exercise 14."""

  def __init__(self, data_path: Text="../data/ch3_q14.csv",
               figsize: Tuple[int, int]=(10, 10)):
    """Constructor

    Attrubutes:
      sim_data: A pd.DataFrame containing the simulation data to analyze.
      figsize: A tuple specifying the default size for figures.
    """
    self.data_path = data_path
    assert os.path.exists(self.data_path), "ch3_q14 dataset not found."
    logging.info("\tLoading %s", self.data_path)
    self.sim_data = pd.read_csv(self.data_path)
    self.figsize = figsize

  def part_b1(self) -> float:
    """
    Using the pearsonr() function from the scipy.stats module, estimate the
    Pearson correlation coefficient.

    Returns:
      A float value containing the Pearson correlation coefficient.
    """
    # YOUR CODE STARTS HERE
    corr, _ = pearsonr(self.sim_data.x1, self.sim_data.x2)
    # raise NotImplementedError
    # YOUR CODE ENDS HERE
    return corr

  def part_b2(self) -> matplotlib.collections.PathCollection:
    """
    Use the scatter() method from matplotlib to return a scatter plot of
    x1 vs x2. Comment on what you see.

    Returns:
      A matplotlib.collections.PathCollection object containing the scatter plot.
    """
    # YOUR CODE STARTS HERE
    _, ax = plt.subplots(figsize=self.figsize)
    ax.set_xlabel("x2")
    ax.set_ylabel("x1")
    figure = ax.scatter(self.sim_data.x2, self.sim_data.x1)
    # raise NotImplementedError
    # YOUR CODE ENDS HERE
    return figure

  def part_c(self) -> statsmodels.iolib.summary.Summary:
    """
    Use the statsmodels ols() function to perform a multiple linear regression with y
    as the response and x1, x2 as the predictors. Return the results using the
    summary() method. Comment on the output.

    Returns:
      A statsmodels Summary object, containing the summary results of the ols
      fit.
    """
    # YOUR CODE STARTS HERE
    formula = ("y ~ x1 + x2")
    ols_fit = ols(formula, data=self.sim_data).fit()
    ols_summary = ols_fit.summary()
    # YOUR CODE ENDS HERE
    return ols_summary

  def part_d(self) -> statsmodels.iolib.summary.Summary:
    """
    Use the sm.OLS() function to perform a multiple linear regression with y
    as the response and x1 as the predictor. Return the results using the
    summary() method. Comment on the output.
    """
    # YOUR CODE STARTS HERE
    formula = ("y ~ x1")
    ols_fit = ols(formula, data=self.sim_data).fit()
    ols_summary = ols_fit.summary()
    # YOUR CODE ENDS HERE
    return ols_summary

  def part_e(self) -> statsmodels.iolib.summary.Summary:
    """
    Use the statsmodels ols() function to perform a multiple linear regression with y
    as the response and x2 as the predictor. Return the results using the
    summary() method. Comment on the output.
    """
    # YOUR CODE STARTS HERE
    formula = ("y ~ x2")
    ols_fit = ols(formula, data=self.sim_data).fit()
    ols_summary = ols_fit.summary()
    # YOUR CODE ENDS HERE
    return ols_summary

  def part_g1(self) -> statsmodels.iolib.summary.Summary:
    """
    Use the statsmodels ols() function to perform a multiple linear regression with y
    as the response and x1, x2 as the predictors on the updated data. Return
    the results using the summary() method. Comment on the output.

    Returns:
      A statsmodels Summary object, containing the summary results of the ols
      fit.
    """
    extra_obs = pd.DataFrame({"x1": [0.1], "x2":[0.8], "y": [6]})
    updated_sim_data = self.sim_data.append(extra_obs)
    # YOUR CODE STARTS HERE
    formula = ("y ~ x1 + x2")
    ols_fit = ols(formula, data=updated_sim_data).fit()
    ols_summary = ols_fit.summary()
    # YOUR CODE ENDS HERE
    return ols_summary

  def part_g2(self) -> statsmodels.iolib.summary.Summary:
    """
    Use the sm.OLS() function to perform a multiple linear regression with y
    as the response and x1 as the predictors on the updated data. Return the
    results using the summary() method. Comment on the output.

    Returns:
      A statsmodels Summary object, containing the summary results of the ols
      fit.
    """
    extra_obs = pd.DataFrame({"x1": [0.1], "x2":[0.8], "y": [6]})
    updated_sim_data = self.sim_data.append(extra_obs)
    # YOUR CODE STARTS HERE
    formula = ("y ~ x1")
    ols_fit = ols(formula, data=updated_sim_data).fit()
    ols_summary = ols_fit.summary()
    # YOUR CODE ENDS HERE
    return ols_summary

  def part_g3(self) -> statsmodels.iolib.summary.Summary:
    """
    Use the statsmodels ols() function to perform a multiple linear regression with y
    as the response and x2 as the predictors on the updated data. Return the
    results using the summary() method. Comment on the output.

    Returns:
      A statsmodels Summary object, containing the summary results of the ols
      fit.
    """
    extra_obs = pd.DataFrame({"x1": [0.1], "x2":[0.8], "y": [6]})
    updated_sim_data = self.sim_data.append(extra_obs)
    # YOUR CODE STARTS HERE
    formula = ("y ~ x2")
    ols_fit = ols(formula, data=updated_sim_data).fit()
    ols_summary = ols_fit.summary()
    # YOUR CODE ENDS HERE
    return ols_summary

def ch3_ex14():
  """Wrapper function for Chapter 3, Exercise 14."""
  logging.info("Chapter 3, Exercise 14.")
  sim = Simulation()

  logging.info("\tPart b.")
  corr = sim.part_b1()
  with open("../results/python/ch3_ex14_part_b1.csv", "w") as f:
    f.write(str(corr))

  sim.part_b2()
  plt.savefig("../results/python/ch3_ex14_part_b2.png")

  logging.info("\tPart c.")
  ols = sim.part_c()
  with open("../results/python/ch3_ex14_part_c.csv", "w") as f:
    f.write(ols.as_text())

  logging.info("\tPart d.")
  ols = sim.part_d()
  with open("../results/python/ch3_ex14_part_d.csv", "w") as f:
    f.write(ols.as_text())

  logging.info("\tPart e.")
  ols = sim.part_e()
  with open("../results/python/ch3_ex14_part_e.csv", "w") as f:
    f.write(ols.as_text())

  logging.info("\tPart g.")
  ols = sim.part_g1()
  with open("../results/python/ch3_ex14_part_g1.csv", "w") as f:
    f.write(ols.as_text())
  ols = sim.part_g2()
  with open("../results/python/ch3_ex14_part_g2.csv", "w") as f:
    f.write(ols.as_text())
  ols = sim.part_g3()
  with open("../results/python/ch3_ex14_part_g3.csv", "w") as f:
    f.write(ols.as_text())

